packages/@webex/plugin-meetings/src/media/MediaConnectionAwaiter.ts

```diff
@@ -0,0 +1,178 @@
+import {Defer} from '@webex/common';
+import {ConnectionState, Event} from '@webex/internal-media-core';
+import LoggerProxy from '../common/logs/logger-proxy';
+import {ICE_AND_DTLS_CONNECTION_TIMEOUT} from '../constants';
+
+export interface MediaConnectionAwaiterProps {
+  webrtcMediaConnection: any;
+}
+
+/**
+ * @class MediaConnectionAwaiter
+ */
+export default class MediaConnectionAwaiter {
+  private webrtcMediaConnection: any;
+  private timer: any;
+  private defer: Defer;
+  private retried: boolean;
+  private onTimeoutCallback: () => void;
+  private connectionStateCallback: () => void;
+  private iceGatheringStateCallback: () => void;
+
+  /**
+   * @param {MediaConnectionAwaiterProps} mediaConnectionAwaiterProps
+   */
+  constructor({webrtcMediaConnection}: MediaConnectionAwaiterProps) {
+    this.webrtcMediaConnection = webrtcMediaConnection;
+    this.defer = new Defer();
+    this.retried = false;
+    this.onTimeoutCallback = this.onTimeout.bind(this);
+    this.connectionStateCallback = this.connectionStateListenerCallback.bind(this);
+    this.iceGatheringStateCallback = this.iceGatheringStateListenerCallback.bind(this);
+  }
+
+  /**
+   * Returns true if the connection is connected, false otherwise.
+   *
+   * @returns {boolean}
+   */
+  private isConnected(): boolean {
+    return this.webrtcMediaConnection.getConnectionState() === ConnectionState.Connected;
+  }
+
+  /**
+   * Returns true if the ICE Gathering is completed, false otherwise.
+   *
+   * @returns {boolean}
+   */
+  private isIceGatheringCompleted(): boolean {
+    return this.webrtcMediaConnection.getIceGatheringState() === 'complete';
+  }
+
+  /**
+   * Clears the callbacks.
+   *
+   * @returns {void}
+   */
+  private clearCallbacks(): void {
+    this.webrtcMediaConnection.off(
+      Event.ICE_GATHERING_STATE_CHANGED,
+      this.iceGatheringStateCallback
+    );
+    this.webrtcMediaConnection.off(Event.CONNECTION_STATE_CHANGED, this.connectionStateCallback);
+
+    this.onTimeoutCallback = null;
+    this.connectionStateCallback = null;
```

Comment:
maybe it's better if we don't reset `connectionStateCallback` and `iceGatheringStateCallback` here, because they're only set in constructor, so the waitForMediaConnectionConnected() won't work if it's called a 2nd time later. Although probably it won't work correctly anyway because of other state like `retried`... so maybe not a big deal, but at least worth adding a comment somewhere that `waitForMediaConnectionConnected()` can only be used once?

Reply:
The question is, if it's fine not to clear it here at all. I believe that should be deleted anyway by garbage collector? So maybe it would be fine to remove those two lines.

Reply:
Done